aws/resource_aws_servicecatalog_portfolio_principal_association.go

```diff
@@ -0,0 +1,165 @@
+		PortfolioId:   aws.String(portfolioId),
+		PrincipalARN:  aws.String(principalArn),
+		PrincipalType: aws.String("IAM"),
+	}
+	conn := meta.(*AWSClient).scconn
+	_, err := conn.AssociatePrincipalWithPortfolio(&input)
+	if err != nil {
+		return fmt.Errorf("creating Service Catalog Principal(%s)/Portfolio(%s) Association failed: %s",
+			principalArn, portfolioId, err.Error())
+	}
+
+	stateConf := resource.StateChangeConf{
+		Pending:      []string{servicecatalog.StatusCreating},
+		Target:       []string{servicecatalog.StatusAvailable},
+		Timeout:      1 * time.Minute,
+		PollInterval: 3 * time.Second,
+		Refresh: func() (interface{}, string, error) {
+			err := resourceAwsServiceCatalogPortfolioPrincipalAssociationRead(d, meta)
+			if err != nil {
+				return 42, "", err
+			}
+			if d.Id() != "" {
+				return 42, servicecatalog.StatusAvailable, err
+			}
+			return 0, servicecatalog.StatusCreating, err
+		},
+	}
+	_, err = stateConf.WaitForState()
+	return err
+}
+
+func resourceAwsServiceCatalogPortfolioPrincipalAssociationRead(d *schema.ResourceData, meta interface{}) error {
+	id, portfolioId, principalArn := resourceAwsServiceCatalogPortfolioPrincipalAssociationRequiredParameters(d)
+	input := servicecatalog.ListPrincipalsForPortfolioInput{
+		PortfolioId: aws.String(portfolioId),
+	}
+	conn := meta.(*AWSClient).scconn
+	var pageToken = ""
+	isFound := false
+	for {
+		pageOfDetails, nextPageToken, err := resourceAwsServiceCatalogPortfolioPrincipalAssociationListPrincipalsForPortfolioPage(conn, input, &pageToken)
+		if err != nil {
+			return err
+		}
+		for _, principal := range pageOfDetails {
+			if *principal.PrincipalARN == principalArn {
+				isFound = true
+				d.SetId(id)
+				break
+			}
+		}
+		if nextPageToken == nil || isFound {
+			break
+		}
+		pageToken = *nextPageToken
+	}
+	if !isFound {
+		log.Printf("[WARN] Service Catalog Principal(%s)/Portfolio(%s) Association not found, removing from state",
+			principalArn, portfolioId)
+		d.SetId("")
+	}
+	d.Set("principal_arn", principalArn)
+	d.Set("portfolio_id", portfolioId)
+	return nil
+}
+
+func resourceAwsServiceCatalogPortfolioPrincipalAssociationListPrincipalsForPortfolioPage(conn *servicecatalog.ServiceCatalog, input servicecatalog.ListPrincipalsForPortfolioInput, nextPageToken *string) ([]*servicecatalog.Principal, *string, error) {
+	input.PageToken = nextPageToken
+	var page, err = conn.ListPrincipalsForPortfolio(&input)
+	if err != nil {
+		return nil, nil, fmt.Errorf("retrieving Service Catalog Associations for Principal/Portfolios: %s", err.Error())
+	}
+	principalDetails := page.Principals
+	return principalDetails, page.NextPageToken, nil
+}
+
+func resourceAwsServiceCatalogPortfolioPrincipalAssociationUpdate(d *schema.ResourceData, meta interface{}) error {
+	const principalArnKey = "principal_arn"
+	const portfolioIdKey = "portfolio_id"
+	if d.HasChange(principalArnKey) || d.HasChange(portfolioIdKey) {
```

Comment:
change to a single `d.HasChanges(principalArnKey, portfolioIdKey)`

Reply:
btw, any reason for the consts? using the string literals are the way its done in other resources

Reply:
the linter picked these up and already changed (already checked in)

the only reason for the constants was that in some projects that's preferred (to prevent typos) but we've switched to this project's convention of repreated string literals